Add tracer interpolation method for missing levels when reading grib2 input

JiliDong-NOAA opened the pull request:
A new method ("intrp") is added for tracers to interpolate to the missing isobaric levels for grib2 input. This pressure to pressure interpolation is mostly to handle the combined GFSv16 grib2 files where specific humidity has missing levels all through the depth and assigning a small constant value (e.g. 1E-7) in the whole column is not always desirable particularly in mid to lower troposphere.

The interpolation is a simple linear one in log(P) coordinate.

To use the option, just replace spfh method from "set_to_fill" to "intrp" in varmap. As this is a new option, it is not expected to change/break any current existing applications if "intrp" is not used.  

Fixes #555

JiliDong-NOAA commented:
This is related to issue #555

JiliDong-NOAA commented:
Not sure how to solve this test error:

_/home/runner/work/UFS_UTILS/UFS_UTILS/ufs_utils/sorc/chgres_cube.fd/input_data.F90:6717: error: Member dint2p(PPIN, XXIN, NPIN, PPOUT, XXOUT, NPOUT, LINLOG, XMSG, IER) (function) of module input_data is not documented. (warning treated as error, aborting now)_

dint2p is a function added for pres to pres interpolation. The compiling and testing are good on Hera though.

GeorgeGayno-NOAA commented:
> Not sure how to solve this test error:
> 
> _/home/runner/work/UFS_UTILS/UFS_UTILS/ufs_utils/sorc/chgres_cube.fd/input_data.F90:6717: error: Member dint2p(PPIN, XXIN, NPIN, PPOUT, XXOUT, NPOUT, LINLOG, XMSG, IER) (function) of module input_data is not documented. (warning treated as error, aborting now)_
> 
> dint2p is a function added for pres to pres interpolation. The compiling and testing are good on Hera though.

@JiliDong-NOAA That is a doxygen error. All new routines must be documented with doxygen.

GeorgeGayno-NOAA commented:
@JiliDong-NOAA We should add a unit test for this new routine. I can help with that.

JiliDong-NOAA commented:
> > Not sure how to solve this test error:
> > _/home/runner/work/UFS_UTILS/UFS_UTILS/ufs_utils/sorc/chgres_cube.fd/input_data.F90:6717: error: Member dint2p(PPIN, XXIN, NPIN, PPOUT, XXOUT, NPOUT, LINLOG, XMSG, IER) (function) of module input_data is not documented. (warning treated as error, aborting now)_
> > dint2p is a function added for pres to pres interpolation. The compiling and testing are good on Hera though.
> 
> @JiliDong-NOAA That is a doxygen error. All new routines must be documented with doxygen.

Thanks @GeorgeGayno-NOAA . Doxygen documentation is added to dint2p and all checks have passed.

JiliDong-NOAA commented:
> @JiliDong-NOAA We should add a unit test for this new routine. I can help with that.

@GeorgeGayno-NOAA Please let me know how I should go forward with the unit test as I have no idea how to do that.

GeorgeGayno-NOAA commented:
> > @JiliDong-NOAA We should add a unit test for this new routine. I can help with that.
> 
> @GeorgeGayno-NOAA Please let me know how I should go forward with the unit test as I have no idea how to do that.

I can help with a unit test when I return from lunch. I will create a simple test and check it into your branch. Make sure I can access your fork.

GeorgeGayno-NOAA commented on sorc/chgres_cube.fd/input_data.F90:
```diff
@@ -6681,4 +6718,226 @@ subroutine check_cnwat(cnwat)
+!> caution. The routine is mostly for GFSV16 combined grib2 input when spfh has
+!> missing levels in low and mid troposphere from U/T/HGT/DZDT. 
+!!
+!! @param ppin     [in] 1d input pres levs
```
Check the parameter syntax. It should be done as follows: @param [in] ppin 1d input pres levs

GeorgeGayno-NOAA commented on sorc/chgres_cube.fd/input_data.F90:
```diff
@@ -6681,4 +6718,226 @@ subroutine check_cnwat(cnwat)
+!! @param linlog   [in] interp method.1:linear;not 1:log;neg:extrp allowed
+!! @param xmsg     [in] fill values of missing levels (-999.0)
+!! @param ier      [out] error status. non 0: failed interpolation
+!! @author NCL code for pressure level interpolation
```
Remove the first @author. 

Do author as follows: @author Jili Dong NCEP/EMC

GeorgeGayno-NOAA replied:
I would add a @date as well.

GeorgeGayno-NOAA replied:
Sorry, I meant lowercase: @date. Otherwise you get a doxygen warning. You will note that one of the checks failed.

GeorgeGayno-NOAA replied:
One more time: `@date`

JiliDong-NOAA replied:
> One more time: `@date`

changed already. But the intel test failed at downloading some index file:

_E: Some index files failed to download. They have been ignored, or old ones used instead.
Error: Process completed with exit code 100._

Don't know how to rerun the test.

GeorgeGayno-NOAA commented on sorc/chgres_cube.fd/input_data.F90:
```diff
@@ -6681,4 +6718,226 @@ subroutine check_cnwat(cnwat)
+! Hence, the C-Wrapper was used to create 'work' arrays which
+! were then passed to this code.  The original focused (non-NCL) 
+! task was to handle PPIN & PPOUT that had the same 'monotonicity.' 
+! Extra code was added to handle the more general case. 
```
This statement does not give me much confidence.

JiliDong-NOAA replied:
> This statement does not give me much confidence.

Same feeling when I first read this. After looking through the code, I didn't see anything suspicious. Given that this NCL code has been used for a while and is quite simple, it should be ok for our applications here. Hope unit test can also confirm this too.

GeorgeGayno-NOAA replied:
I don't think the failure is anything you did. There is a rerun-jobs button you can click on.

JiliDong-NOAA replied:
> I don't think the failure is anything you did. There is a rerun-jobs button you can click on.

I can't find the re-run button. Or I don't have access/permission to it?

GeorgeGayno-NOAA replied:
Click on the failed test "details". You will see the button at the upper right.

JiliDong-NOAA replied:
> Click on the failed test "details". You will see the button at the upper right.

still can't find it. Please see attached screenshots.

https://drive.google.com/file/d/1wfQbj77Ug4YyJL91Tj3zZrUBisDpC05c/view?usp=sharing
https://drive.google.com/file/d/1U4VcTEiruhTNCi0K-dFYOkikwpeMXtiL/view?usp=sharing

GeorgeGayno-NOAA replied:
Looks like you don't have permissions. I will rerun for you.

JiliDong-NOAA replied:
> Looks like you don't have permissions. I will rerun for you.

Thank you @GeorgeGayno-NOAA Now it all passed.

GeorgeGayno-NOAA replied:
Great. Next I will add a sample unit test to your branch.

GeorgeGayno-NOAA commented:
I added a unit test at c3bf3fb. You will need to add some comments to the source code. Also, you will want to improve and expand on what I did. For example, the routine has different interpolation options - linear, logp, extrapolation. You want to test each option.

JiliDong-NOAA commented:
> I added a unit test at [c3bf3fb](https://github.com/NOAA-EMC/UFS_UTILS/commit/c3bf3fb324825827f44b38b98d2a5b7db3966b81). You will need to add some comments to the source code. Also, you will want to improve and expand on what I did. For example, the routine has different interpolation options - linear, logp, extrapolation. You want to test each option.

Thanks @GeorgeGayno-NOAA I will work on it. How would I run the unit test? Using a single job card for "srun ftst_dint2p"?

GeorgeGayno-NOAA commented on sorc/chgres_cube.fd/input_data.F90:
```diff
@@ -2749,6 +2763,28 @@ subroutine read_input_atm_grib2_file(localpet)
+             dummy3d_col_in=dummy3d(ii,jj,:)
+             call dint2p(rlevs,dummy3d_col_in,lev_input,rlevs,dummy3d_col_out,    &
+                        lev_input, 2, intrp_missing, intrp_ier) 
+             if (intrp_ier .gt. 0) then
```
chgres_cube has an 'official' error handling routine. Call it here (routine error_handler).

GeorgeGayno-NOAA replied:
And below.

JiliDong-NOAA replied:
error_handler added

GeorgeGayno-NOAA commented:
> > I added a unit test at [c3bf3fb](https://github.com/NOAA-EMC/UFS_UTILS/commit/c3bf3fb324825827f44b38b98d2a5b7db3966b81). You will need to add some comments to the source code. Also, you will want to improve and expand on what I did. For example, the routine has different interpolation options - linear, logp, extrapolation. You want to test each option.
> 
> Thanks @GeorgeGayno-NOAA I will work on it. How would I run the unit test? Using a single job card for "srun ftst_dint2p"?

You can run it on Hera, Jet or Orion. First, to build_all.sh, add 'make test' after the 'make install'. Then, under ./cmake, edit the mpiexec file for your machine for an account and queue you are eligible to use. That will automatically run all the tests. Log files from the tests will be under ./build/Testing/Temporary.

GeorgeGayno-NOAA commented:
@JiliDong-NOAA Since you are adding a new option, you will need to update the related documentation:

- https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/parm/varmap_tables/README
- https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/sorc/chgres_cube.fd/docs/user_guide.md
- https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/docs/source/chgres_cube.rst

JiliDong-NOAA commented:
> > > I added a unit test at [c3bf3fb](https://github.com/NOAA-EMC/UFS_UTILS/commit/c3bf3fb324825827f44b38b98d2a5b7db3966b81). You will need to add some comments to the source code. Also, you will want to improve and expand on what I did. For example, the routine has different interpolation options - linear, logp, extrapolation. You want to test each option.
> > 
> > 
> > Thanks @GeorgeGayno-NOAA I will work on it. How would I run the unit test? Using a single job card for "srun ftst_dint2p"?
> 
> You can run it on Hera, Jet or Orion. First, to build_all.sh, add 'make test' after the 'make install'. Then, under ./cmake, edit the mpiexec file for your machine for an account and queue you are eligible to use. That will automatically run all the tests. Log files from the tests will be under ./build/Testing/Temporary.

dint2p unit test successful after a bug is fixed when using lnP extrapolation (the routine won't use extrapolation option when being called but it is safe to get the big fixed).

The unit test consider the four scenarios:

1. linear interpolation; no extrapolation
2. linear interpolation; linear extrapolation
3. lnP interpolation; no extrapolation
4. lnP interpolation; lnP extrapolation

      Start  6: chgres_cube-ftst_dint2p
 6/18 Test  #6: chgres_cube-ftst_dint2p .....................   Passed    0.03 sec


There are several unit tests failed which seem not related to my change:

The following tests FAILED:
	  9 - chgres_cube-ftst_read_sfc_gfs_nemsio (Failed)
	 10 - chgres_cube-ftst_read_sfc_nemsio (Failed)
	 11 - chgres_cube-ftst_read_sfc_netcdf (Failed)
	 12 - chgres_cube-ftst_read_nst_netcdf (Failed)
	 13 - chgres_cube-ftst_read_nst_nemsio (Failed)
	 14 - chgres_cube-ftst_read_atm_gaussian_netcdf (Failed)
	 18 - sfc_climo_gen-ftst_interp (Failed)

GeorgeGayno-NOAA commented:
@JiliDong-NOAA Don't worry about the tests that failed on Hera. You did not break anything.

JiliDong-NOAA commented:
> @JiliDong-NOAA Since you are adding a new option, you will need to update the related documentation:
> 
> * https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/parm/varmap_tables/README
> * https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/sorc/chgres_cube.fd/docs/user_guide.md
> * https://github.com/JiliDong-NOAA/UFS_UTILS/blob/add_intrp/docs/source/chgres_cube.rst

related documentation updated

GeorgeGayno-NOAA commented:
@JiliDong-NOAA Did you run the chgres consistency tests?

JiliDong-NOAA commented:
> @JiliDong-NOAA Did you run the chgres consistency tests?

@GeorgeGayno-NOAA I did it on Hera and they all passed:

consistency.log01:<<< C96 FV3 RESTART TEST PASSED. >>>
consistency.log02:<<< C192 FV3 HISTORY TEST PASSED. >>>
consistency.log03:<<< C96 FV3 GAUSSIAN NEMSIO TEST PASSED. >>>
consistency.log04:<<< C96 GFS SIGIO TEST PASSED. >>>
consistency.log05:<<< C96 GFS GAUSSIAN NEMSIO TEST PASSED. >>>
consistency.log06:<<< C96 REGIONAL TEST PASSED. >>>
consistency.log07:<<< C96 FV3 GAUSSIAN NETCDF TEST PASSED. >>>
consistency.log08:<<< C192 GFS GRIB2 TEST PASSED. >>>
consistency.log09:<<< 25-KM CONUS GFS GRIB2 TEST PASSED. >>>
consistency.log10:<<< 3-km CONUS HRRR W/ GFS PHYSICS GRIB2 TEST PASSED. >>>
consistency.log11:<<< 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
consistency.log12:<<< 13-KM CONUS NAM GRIB2 TEST PASSED. >>>
consistency.log13:<<< 13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
consistency.log14:<<< 13-KM NA GFS NCEI GRIB2 TEST PASSED. >>>
consistency.log15:<<< C96 FV3 GAUSSIAN NETCDF2WAM TEST PASSED. >>>


Do I also need to run the consistency tests on other platforms?

GeorgeGayno-NOAA commented:
> > @JiliDong-NOAA Did you run the chgres consistency tests?
> 
> @GeorgeGayno-NOAA I did it on Hera and they all passed:
> 
> consistency.log01:<<< C96 FV3 RESTART TEST PASSED. >>>
> consistency.log02:<<< C192 FV3 HISTORY TEST PASSED. >>>
> consistency.log03:<<< C96 FV3 GAUSSIAN NEMSIO TEST PASSED. >>>
> consistency.log04:<<< C96 GFS SIGIO TEST PASSED. >>>
> consistency.log05:<<< C96 GFS GAUSSIAN NEMSIO TEST PASSED. >>>
> consistency.log06:<<< C96 REGIONAL TEST PASSED. >>>
> consistency.log07:<<< C96 FV3 GAUSSIAN NETCDF TEST PASSED. >>>
> consistency.log08:<<< C192 GFS GRIB2 TEST PASSED. >>>
> consistency.log09:<<< 25-KM CONUS GFS GRIB2 TEST PASSED. >>>
> consistency.log10:<<< 3-km CONUS HRRR W/ GFS PHYSICS GRIB2 TEST PASSED. >>>
> consistency.log11:<<< 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
> consistency.log12:<<< 13-KM CONUS NAM GRIB2 TEST PASSED. >>>
> consistency.log13:<<< 13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
> consistency.log14:<<< 13-KM NA GFS NCEI GRIB2 TEST PASSED. >>>
> consistency.log15:<<< C96 FV3 GAUSSIAN NETCDF2WAM TEST PASSED. >>>
> 
> Do I also need to run the consistency tests on other platforms?

We need to run on WCOSS-Dell, WCOSS-Cray, Jet and Orion. If you don't have access to a machine let me know and I will run the test for you.

JiliDong-NOAA commented:
> > > @JiliDong-NOAA Did you run the chgres consistency tests?
> > 
> > 
> > @GeorgeGayno-NOAA I did it on Hera and they all passed:
> > consistency.log01:<<< C96 FV3 RESTART TEST PASSED. >>>
> > consistency.log02:<<< C192 FV3 HISTORY TEST PASSED. >>>
> > consistency.log03:<<< C96 FV3 GAUSSIAN NEMSIO TEST PASSED. >>>
> > consistency.log04:<<< C96 GFS SIGIO TEST PASSED. >>>
> > consistency.log05:<<< C96 GFS GAUSSIAN NEMSIO TEST PASSED. >>>
> > consistency.log06:<<< C96 REGIONAL TEST PASSED. >>>
> > consistency.log07:<<< C96 FV3 GAUSSIAN NETCDF TEST PASSED. >>>
> > consistency.log08:<<< C192 GFS GRIB2 TEST PASSED. >>>
> > consistency.log09:<<< 25-KM CONUS GFS GRIB2 TEST PASSED. >>>
> > consistency.log10:<<< 3-km CONUS HRRR W/ GFS PHYSICS GRIB2 TEST PASSED. >>>
> > consistency.log11:<<< 3-km CONUS HRRR W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
> > consistency.log12:<<< 13-KM CONUS NAM GRIB2 TEST PASSED. >>>
> > consistency.log13:<<< 13-km CONUS RAP W/ GSD PHYSICS AND SFC FROM FILE GRIB2 TEST PASSED. >>>
> > consistency.log14:<<< 13-KM NA GFS NCEI GRIB2 TEST PASSED. >>>
> > consistency.log15:<<< C96 FV3 GAUSSIAN NETCDF2WAM TEST PASSED. >>>
> > Do I also need to run the consistency tests on other platforms?
> 
> We need to run on WCOSS-Dell, WCOSS-Cray, Jet and Orion. If you don't have access to a machine let me know and I will run the test for you.

@GeorgeGayno-NOAA consistency tests all passed on Hera, Jet, Orion, WCOSS-Cray and WCOSS-Dell

GeorgeGayno-NOAA commented:
@LarissaReames-NOAA and @JeffBeck-NOAA You know the grib2 option better than I do. I would like one of you to review this PR. Thanks.

JeffBeck-NOAA commented:
> Hi @JiliDong-NOAA, based on the issue conversion, there was discussion of adding a check that will cause chgres_cube to fail if 1) Interp isn't selected, and 2) levels of a tracer are missing below 200 mb. We need to figure out how to incorporate that change into this PR, but also making sure it won't fail for older GFS files that just don't contain certain fields. It needs to differentiate between whether levels are missing (and fail in that instance if "interp" isn't selected), but also allow the code to continue if it's just a field that simply doesn't exist in that external model data. @LarissaReames-NOAA has a good idea on how to implement this check, so I'll let her chime in.
> 
> sounds good @JeffBeck-NOAA . I am opening access of the branch add_intrp to both you and @LarissaReames-NOAA . Feel free to make any changes. Or let me know any modifications you propose and I am happy to implement them in this PR.

Moving this over to the PR instead.  @JiliDong-NOAA, thanks.  I'll let @LarissaReames-NOAA respond, then we'll see how to implement this check.

LarissaReames-NOAA commented:
> > Hi @JiliDong-NOAA, based on the issue conversion, there was discussion of adding a check that will cause chgres_cube to fail if 1) Interp isn't selected, and 2) levels of a tracer are missing below 200 mb. We need to figure out how to incorporate that change into this PR, but also making sure it won't fail for older GFS files that just don't contain certain fields. It needs to differentiate between whether levels are missing (and fail in that instance if "interp" isn't selected), but also allow the code to continue if it's just a field that simply doesn't exist in that external model data. @LarissaReames-NOAA has a good idea on how to implement this check, so I'll let her chime in.
> > sounds good @JeffBeck-NOAA . I am opening access of the branch add_intrp to both you and @LarissaReames-NOAA . Feel free to make any changes. Or let me know any modifications you propose and I am happy to implement them in this PR.
> 
> Moving this over to the PR instead. @JiliDong-NOAA, thanks. I'll let @LarissaReames-NOAA respond, then we'll see how to implement this check.

Yeah I don't think this will be super difficult, luckily, we'll just need to add a bit more logic, and I might add another consistency test with an older GFS file to double check that we're not breaking that. I don't think we will be, as I've double checked some rather old data from NCEI and all data are on concurrent levels below 150 mb, so 200 mb should be a safe place to start checking below. 

Basically, the idea is that if there are literally no tracer entries on isobaric (or native) levels, you'd want to just allow the chosen fill/skip behavior to remain and move on, but if you have non-zero number of levels of said tracer on those surfaces, but are still finding missing levels below 200mb, and interp isn't chosen, you'd want to abort with an error message. 

You can use the same method as used in this line that counts the number of vertical levels to begin with in this process. For isobaric data that would be:
`iret = grb2_inq(the_file,inv_file,":UGRD:",lvl_str_space)`
and for hybrid/native level data that would be:
`iret = grb2_inq(the_file,inv_file,":UGRD:",lvl_str_space, " level:")`
where `iret` will return the number of entries that match that description. Of course you'd change `UGRD` to whatever tracer is currently being read in. Essentially here you're searching for all levels instead of a particular level, and this could be done once for each tracer and the result could be saved in a logic variable like `all_missing` or something before the individual levels are searched for. 

Let me know if that makes sense.

JiliDong-NOAA commented:
> > > Hi @JiliDong-NOAA, based on the issue conversion, there was discussion of adding a check that will cause chgres_cube to fail if 1) Interp isn't selected, and 2) levels of a tracer are missing below 200 mb. We need to figure out how to incorporate that change into this PR, but also making sure it won't fail for older GFS files that just don't contain certain fields. It needs to differentiate between whether levels are missing (and fail in that instance if "interp" isn't selected), but also allow the code to continue if it's just a field that simply doesn't exist in that external model data. @LarissaReames-NOAA has a good idea on how to implement this check, so I'll let her chime in.
> > > sounds good @JeffBeck-NOAA . I am opening access of the branch add_intrp to both you and @LarissaReames-NOAA . Feel free to make any changes. Or let me know any modifications you propose and I am happy to implement them in this PR.
> > 
> > 
> > Moving this over to the PR instead. @JiliDong-NOAA, thanks. I'll let @LarissaReames-NOAA respond, then we'll see how to implement this check.
> 
> Yeah I don't think this will be super difficult, luckily, we'll just need to add a bit more logic, and I might add another consistency test with an older GFS file to double check that we're not breaking that. I don't think we will be, as I've double checked some rather old data from NCEI and all data are on concurrent levels below 150 mb, so 200 mb should be a safe place to start checking below.
> 
> Basically, the idea is that if there are literally no tracer entries on isobaric (or native) levels, you'd want to just allow the chosen fill/skip behavior to remain and move on, but if you have non-zero number of levels of said tracer on those surfaces, but are still finding missing levels below 200mb, and interp isn't chosen, you'd want to abort with an error message.
> 
> You can use the same method as used in this line that counts the number of vertical levels to begin with in this process. For isobaric data that would be:
> `iret = grb2_inq(the_file,inv_file,":UGRD:",lvl_str_space)`
> and for hybrid/native level data that would be:
> `iret = grb2_inq(the_file,inv_file,":UGRD:",lvl_str_space, " level:")`
> where `iret` will return the number of entries that match that description. Of course you'd change `UGRD` to whatever tracer is currently being read in. Essentially here you're searching for all levels instead of a particular level, and this could be done once for each tracer and the result could be saved in a logic variable like `all_missing` or something before the individual levels are searched for.
> 
> Let me know if that makes sense.
> 
> EDIT: I figured I'd clarify that it would make the most sense to just go ahead and place this sort of logic inside the loop that reads the tracers, that way you save extra code waste.

@LarissaReames-NOAA Thanks for the discussions! I can't say I fully understand this approach. I add a simple logic here instead when the pressure level is missing in the loop for any tracer:
1. if "intrp" is set, do the interpolation
2. if "intrp" is not set and the level is below 200 mb, program aborts
3. if "intrp" is not set and the level is above 200 mb, fill in as before

the critical level is set as "lev_no_tr_fill" and can be adjusted.
Please feel free to make any changes as needed.

LarissaReames-NOAA commented:
I modified some of Jili's logic so that the intrp option will now work with all tracers in various other file types, including GFS grib2 files that have missing ozone in the lowest 600mb and missing cloud tracers in the top ~70 mb. In these cases, instead of attempting to extrapolate, we simple default back to using the fill value method, as these are the locations that this method was originally supposed to address. It will also now fail with a helpful error message if ozone is missing above 400 mb or humidity/cloud tracers are missing below 200 mb. 

I also added in a new consistency test for the pgrib2+bgrib2 files to make sure that this compatibility isn't broken in the future; driver files were updated accordingly.

All variable mapping tables were also modified to use the intrp option for all tracers. This means that all consistency tests that use GFS data will need to have their baselines updated. 

It appears that we have a really odd failure for one of the GitHub checks that has nothing to do with the code itself. Anyone know what's up with that?

GeorgeGayno-NOAA commented:
> I modified some of Jili's logic so that the intrp option will now work with all tracers in various other file types, including GFS grib2 files that have missing ozone in the lowest 600mb and missing cloud tracers in the top ~70 mb. In these cases, instead of attempting to extrapolate, we simple default back to using the fill value method, as these are the locations that this method was originally supposed to address. It will also now fail with a helpful error message if ozone exists _somewhere_ but is missing above 400 mb or humidity/cloud tracers exist _somewhere_ but are missing below 200 mb and the intrp option is not chosen.
> 
> I also added in a new consistency test for the pgrib2+bgrib2 files to make sure that this compatibility isn't broken in the future; driver files were updated accordingly.
> 
> All variable mapping tables were also modified to use the intrp option for all tracers. This means that all consistency tests that use GFS grib2 data will need to have their baselines updated.
> 
> It appears that we have a really odd failure for one of the GitHub checks that has nothing to do with the code itself. Anyone know what's up with that?

It failed installing the intel compiler. That happens sometimes. I reran the test and it passed. There is a rerun button when you click on tests. As an 'admin' you should have privileges to use it.

JiliDong-NOAA commented:
> I modified some of Jili's logic so that the intrp option will now work with all tracers in various other file types, including GFS grib2 files that have missing ozone in the lowest 600mb and missing cloud tracers in the top ~70 mb. In these cases, instead of attempting to extrapolate, we simple default back to using the fill value method, as these are the locations that this method was originally supposed to address. It will also now fail with a helpful error message if ozone exists _somewhere_ but is missing above 400 mb or humidity/cloud tracers exist _somewhere_ but are missing below 200 mb and the intrp option is not chosen.
> 
> I also added in a new consistency test for the pgrib2+bgrib2 files to make sure that this compatibility isn't broken in the future; driver files were updated accordingly.
> 
> All variable mapping tables were also modified to use the intrp option for all tracers. This means that all consistency tests that use GFS grib2 data will need to have their baselines updated.
> 
> It appears that we have a really odd failure for one of the GitHub checks that has nothing to do with the code itself. Anyone know what's up with that?

Thanks @LarissaReames-NOAA ! It looks good. Just have one minor suggestion. Could we add one more constraint when method is "intrp" and the target pres is out of the source range? e.g. in the above scenario, chgres will fill in with default value only when pres level is above 200 mb. If pres level is below 200 mb, we don't want to fill in 0 values and will just abort chgres. The additional constraint will look like:

after interpolation:

             _do vlev = 1, lev_input
               if ((dummy3d_col_out(vlev) .eq. intrp_missing)) then
                if (rlevs(vlev) .gt. lev_no_tr_fill) then
                  call error_handler("Pres out of range and no extrapolation/filling in happened.",1)
                else 
                  dummy3d_col_out(vlev) = value
                endif
               endif
              end do_

LarissaReames-NOAA commented:
> Thanks @LarissaReames-NOAA ! It looks good. Just have one minor suggestion. Could we add one more constraint when method is "intrp" and the target pres is out of the source range? e.g. in the above scenario, chgres will fill in with default value only when pres level is above 200 mb. If pres level is below 200 mb, we don't want to fill in 0 values and will just abort chgres. The additional constraint will look like:
> 
> after interpolation:
> 
> ```
>          _do vlev = 1, lev_input
>            if ((dummy3d_col_out(vlev) .eq. intrp_missing)) then
>             if (rlevs(vlev) .gt. lev_no_tr_fill) then
>               call error_handler("Pres out of range and no extrapolation/filling in happened.",1)
>             else 
>               dummy3d_col_out(vlev) = value
>             endif
>            endif
>           end do_
> ```

I'll have to modify this some, as ozone actually is expected to be missing in the lowest 600mb while the other tracers are missing above 200 mb, but yeah I'll add this in.

JeffBeck-NOAA commented:
@JiliDong-NOAA, I believe this PR is now complete.  If you're OK, we can rerun the consistency checks one last time, then assuming the pass, we can merge.

JiliDong-NOAA commented:
I made one minor change to move zeroing out tracer to the end of filling in default value. Otherwise it may zero out pre-assigned missing value of -999 before we use it to determine whether to fill in. @LarissaReames-NOAA could you review it? Other than that, it looks good to me.

LarissaReames-NOAA commented:
> I made one minor change to move zeroing out tracer to the end of filling in default value. Otherwise it may zero out pre-assigned missing value of -999 before we use it to determine whether to fill in. @LarissaReames-NOAA could you review it? Other than that, it looks good to me.

Ah, yep, great catch. That looks good to me.

LarissaReames-NOAA commented:
> > > I made one minor change to move zeroing out tracer to the end of filling in default value. Otherwise it may zero out pre-assigned missing value of -999 before we use it to determine whether to fill in. @LarissaReames-NOAA could you review it? Other than that, it looks good to me.
> > 
> > 
> > Ah, yep, great catch. That looks good to me.
> 
> Thanks @LarissaReames-NOAA. I am doing consistency tests on Hera but unable to get consistency.log and summary.log. The individual log files (consistency.log[01-16]) seem ok. Anything I need to change to get those two summary log files?

That's odd. The summary file should be produced so long as all the individual tests complete successfully. I usually only fail to get a summary file when a test fails before it can test for file consistency, so the chgres_cube code fails during execution. I just ran the tests on the most recent code and I get the summary file just fine. You can see the output here: `/scratch1/BMC/gsd-fv3/Larissa.Reames/UFS_UTILS_jili/reg_tests/chgres_cube`

JiliDong-NOAA commented:
> > > > I made one minor change to move zeroing out tracer to the end of filling in default value. Otherwise it may zero out pre-assigned missing value of -999 before we use it to determine whether to fill in. @LarissaReames-NOAA could you review it? Other than that, it looks good to me.
> > > 
> > > 
> > > Ah, yep, great catch. That looks good to me.
> > 
> > 
> > Thanks @LarissaReames-NOAA. I am doing consistency tests on Hera but unable to get consistency.log and summary.log. The individual log files (consistency.log[01-16]) seem ok. Anything I need to change to get those two summary log files?
> 
> That's odd. The summary file should be produced so long as all the individual tests complete successfully. I usually only fail to get a summary file when a test fails before it can test for file consistency, so the chgres_cube code fails during execution. I just ran the tests on the most recent code and I get the summary file just fine. You can see the output here: `/scratch1/BMC/gsd-fv3/Larissa.Reames/UFS_UTILS_jili/reg_tests/chgres_cube`

@LarissaReames-NOAA Now I see where the problem is. I didn't change HOMEreg to /scratch1/BMC/gsd-fv3/Larissa.Reames/chgres_cube/reg_tests. After changing that, I can get the summary.log. Noticed several reg test failures 08, 09 and 14 though. I guess the reg tests will be updated later.

LarissaReames-NOAA commented:
All unit and consistency tests pass on Hera and Jet (other than the expected failures for previous tests using GFS grib2 data). @GeorgeGayno-NOAA can you help run tests on Orion and WCOSS?

GeorgeGayno-NOAA commented:
> All unit and consistency tests pass on Hera and Jet (other than the expected failures for previous tests using GFS grib2 data). @GeorgeGayno-NOAA can you help run tests on Orion and WCOSS?

The tests that failed - will the baseline data need to be updated?

LarissaReames-NOAA commented:
> > All unit and consistency tests pass on Hera and Jet (other than the expected failures for previous tests using GFS grib2 data). @GeorgeGayno-NOAA can you help run tests on Orion and WCOSS?
> 
> The tests that failed - will the baseline data need to be updated?

Yes, there should be three failures -- all the old tests that used GFS grib2 data. They'll need to be updated, and the change is due to the update of all variable mapping tables to interpolate scattered missing levels.

GeorgeGayno-NOAA commented:
I ran the tests on WCOSS-Dell. I got four failures, not three:

```
consistency.log01:<<< 25-KM CONUS GFS GRIB2 TEST FAILED. >>>
consistency.log06:<<< 13-KM NA GFS NCEI GRIB2 TEST FAILED. >>>
consistency.log14:<<< C192 GFS GRIB2 TEST FAILED. >>>
consistency.log16:<<< 25-km CONUS GFS GRIB2 TEST FAILED. <<<
```

LarissaReames-NOAA commented:
> > Ah yes, well that last one will fail if you haven't already created the correct baseline file for the system since it's new in this PR. So the first three will need new baselines, and the last one (test 16) will just need to have whatever file you get from the test itself inserted as the standard baseline result.
> 
> Oh, I forgot you added a new test. Is this the input data for it? `/scratch1/BMC/gsd-fv3/Larissa.Reames/chgres_cube/reg_tests/input_data/gfs.pbgrib2`
Yep, that's the one.

GeorgeGayno-NOAA commented:
@JiliDong-NOAA I found some bugs in some of the regression test scripts. I will check in the fixes to your branch.

GeorgeGayno-NOAA commented:
The consistency tests were run on all Tier 1 machines - Jet, Orion, Hera, WCOSS-Cray and WCOSS-Dell. The baseline data was updated for the three tests that were affected by changes under this PR. And input and baseline data was added for the new consistency test.

GeorgeGayno-NOAA reviewed (approved):
I have no further comments.

GeorgeGayno-NOAA commented:
Are we ready to merge?

JeffBeck-NOAA commented:
> Are we ready to merge?

I believe so.  Thanks for your help with this PR.
